src/js/layer.js

```diff
@@ -0,0 +1,67 @@
+const loginArea = document.querySelector('.header__login');
+const loginLayer = document.querySelector('.header-layer__popup--login');
+const loginExpandLayer = document.querySelector('.header-layer__popup--login-extend');
+const addressArea = document.querySelector('.header__shipping-address');
+const addressLayer = document.querySelector('.header-layer__popup--address');
+const dimBackground = document.querySelector('.dim');
+
+let isOpenAddresslayer = false;
+let isOpenLoginExpandLayer = false;
+
+function renderLoginLayer() {
+  document.addEventListener('DOMContentLoaded', () => {
+    setTimeout(() => {
+      loginLayer.classList.remove('hidden');
+    }, 1000);
+  });
+}
+
+function handleAddressMouseEnter() {
+  if (!isOpenAddresslayer) openLayer(addressLayer);
+  isOpenAddresslayer = true;
+}
+
+function handleLoginMouseEnter() {
+  loginLayer.classList.add('hidden');
+  if (!isOpenAddresslayer) openLayer(loginExpandLayer);
+  isOpenLoginExpandLayer = true;
+}
+
+function handleAddressMouseLeave(e) {
+  isOpenAddresslayer = false;
+  closeLayer(addressLayer, e);
+}
+
+function handleLoginMouseLeave(e) {
+  isOpenLoginExpandLayer = false;
+  closeLayer(loginExpandLayer, e);
+}
+
+function openLayer(modal) {
+  modal.classList.remove('hidden');
+  dimBackground.classList.remove('hidden');
+}
+
+function closeLayer(modal, e) {
+  if (e.relatedTarget !== modal && !isOpenAddresslayer && !isOpenLoginExpandLayer) {
+    modal.classList.add('hidden');
+    dimBackground.classList.add('hidden');
+  }
+}
+
+function addMouseEvents(area, layer, handleMouseEnter, handleMouseLeave) {
+  area.addEventListener('mouseenter', handleMouseEnter);
+  layer.addEventListener('mouseenter', handleMouseEnter);
+  area.addEventListener('mouseleave', handleMouseLeave);
+  layer.addEventListener('mouseleave', handleMouseLeave);
+}
```

Comment:
추가로 작은 의견이 있습니당
여러 요소에서 mouseenter, mouseleave 이벤트를 등록해야 하기 때문에 중복을 제거하기 위해 addMouseEvents 함수로 추출한 것은 이해가 됐는데요!
개인적으로는 각각의 이벤트는 독립적으로 생각이 돼서 한 함수에 묶어놓으니 오히려 헷갈리는 것 같습니다. 🫠

---
여기까지 작성해두고 함수를 다시 보니 area, layer 두개의 파라미터를 받을 필요가 없어 보이네요!
```suggestion
function addMouseEvents(element, handleMouseEnter, handleMouseLeave) {
  element.addEventListener('mouseenter', handleMouseEnter);
  element.addEventListener('mouseleave', handleMouseLeave);
}
```

area와 layer에 대해 같은 동작을 하고 있어서 인자를 하나로 줄이고 호출하는 쪽에서 적절한 인자를 넣어주면 될 것 같습니다!

Reply:
저도 처음 작성할 때 코멘트 해주신 것처럼 위의 코멘트와 아래 코멘트 부분의 중복을 해결해보려고 생각해본게 맞습니다.. 작성하면서도 더 좋은 방법이 있을 것 같은데 하고 생각이 많이 들었어요 😂
위에 코멘트와 함께 적용하려고 고민해봤는데 요건 별로일까요 ??

```
function addMouseEvents(elements, handleMouseEnter, handleMouseLeave) {
  elements.forEach((element) => {
    element.addEventListener('mouseenter', handleMouseEnter);
    element.addEventListener('mouseleave', handleMouseLeave);
  });
}

function initNavLayer() {
  renderLoginLayer();
  addMouseEvents([addressArea, addressLayer], handleAddressMouseEnter, handleAddressMouseLeave);
  addMouseEvents([loginArea, loginExpandLayer], handleLoginMouseEnter, handleLoginMouseLeave);
}
```

Reply:
오호 요렇게 해결할 수도 있네요!
현재의 요구사항에 맞게 중복 로직은 잘 제거가 된 것 같아요.

근데 개인적인 생각으로는 `addMouseEvents` 함수에 mouseenter, mouseleave 이벤트를 함께 처리하도록 하는 것보다
각각의 이벤트를 직접 등록해주는 게 더 좋을 것 같다는 생각이 들어요!

가장 큰 이유는 지금 요구사항에서는 mouseenter, mouseleave 할 때의 동작으로 기능명세가 되어 있지만
만약에 mouseenter에 모달 요소를 보여주고 모달을 끄는 건 모달 내의 버튼을 클릭했을 때로 요구사항이 바뀐다면?
그때는 addMouseEvents 를 재사용하지 못하게 되겠죠 😔
물론 이렇게 요구사항이 바뀔 때 addMouseEvents에 handleMouseClick 이라는 파라미터를 추가해서 해결을 할 수도 있지만
만약 모달 요소 안에 여러가지 버튼이 있고 그 버튼마다 클릭했을 때 동작이 다르다면??
handleMouseClick1, handleMouseClick2, ... 이렇게 파라미터를 늘려가는 것보다는
한 요소에 이벤트 핸들러를 하나씩 (별도의 함수로 분리하지 않고) 처리해주는 게
지금 당장은 중복처럼 보이지만 이정도의 중복은 괜찮고, 나중에 수정하기도 편하지 않을까 하는 개인적인 의견이었습니다. 😆

제 의견이 정답은 아니니 한번 고려해보시고, 아티가 맞다고 생각하는 방법으로 하시면 됩니다 !!

```js
addressArea.addEventListener('mouseenter', handleAddressMouseEnter);
addressArea.addEventListener('mouseleave', handleAddressMouseLeave);
...
```

Reply:
너무 줄여보려고 했었나 봅니다.. 감사합니다

Reply:
적정선을 찾기가 참 어려운 것 같아요. 저도 일하면서 같은 내용으로 리뷰를 많이 받기도 하고 항상 논쟁도 있고 그렇습니당 ㅠㅠ ㅋㅋㅋㅋ